feat: Add shutdown handler for client to call and close alive sockets

varunverlencar opened the pull request:
Add TCP port socket shutdown if not alive

**Public API Changes**
<!-- Describe any changes to the public API, or write "None" -->
Add shutdown handler for clients **/shutdown**

**Description**
<!-- Describe what has changed, and motivation behind those changes -->

1. Added changes for shutdown handler and api
2. Added readme usage instructions
3. tested on local javascipt UI


<!-- Link relevant GitHub issues -->
https://github.com/RobotWebTools/web_video_server/issues/193

varunverlencar commented:
tested with `ros2 run image_tools cam2image --ros-args -p frequency:=30.0 -p width:=640 -p height:=480 -p burger_mode:=true`
and custom UI that listens on web_video_server: you can also use web-browser `http://localhost:8080/shutdown?topic=/image`
output: 
```
[INFO] [web_video_server-1]: process started with pid [246966]
[web_video_server-1] [INFO] [1763087236.891895946] [web_video_server]: Waiting For connections on 0.0.0.0:8080
[web_video_server-1] [INFO] [1763087282.546730930] [web_video_server]: Handling Request: /stream?topic=/image
[web_video_server-1] [INFO] [1763087282.547466223] [web_video_server]: Streaming topic /image with QoS profile default
[web_video_server-1] [INFO] [1763087290.387515416] [web_video_server]: Handling Request: /shutdown
[web_video_server-1] [INFO] [1763087290.387721288] [web_video_server]: Shutdown request removed 1 stream(s) for topic <all>
[web_video_server-1] [INFO] [1763087290.388883366] [web_video_server]: Handling Request: /shutdown
[web_video_server-1] [INFO] [1763087290.388922029] [web_video_server]: Shutdown request removed 0 stream(s) for topic <all>
[web_video_server-1] [INFO] [1763087290.393034420] [web_video_server]: Removed Stream: /image
[web_video_server-1] [INFO] [1763087296.019929017] [web_video_server]: Handling Request: /stream?topic=/image
[web_video_server-1] [INFO] [1763087296.020282519] [web_video_server]: Streaming topic /image with QoS profile default
[web_video_server-1] [INFO] [1763087299.776973402] [web_video_server]: Handling Request: /shutdown
[web_video_server-1] [INFO] [1763087299.777058013] [web_video_server]: Shutdown request removed 1 stream(s) for topic <all>
[web_video_server-1] [INFO] [1763087299.778284809] [web_video_server]: Handling Request: /shutdown
[web_video_server-1] [INFO] [1763087299.778319854] [web_video_server]: Shutdown request removed 0 stream(s) for topic <all>
[web_video_server-1] [INFO] [1763087299.892136295] [web_video_server]: Removed Stream: /image
```

bjsowa commented:
Sorry for the lack of response. This looks overall good but I recently refactored the project and changed the structure a lot so the changes can't be applied as is. Could you resolve the conflicts or just rewrite the PR for the new structure?

varunverlencar committed:
feat: add /shutdown endpoint to stop active streams (resolves #193)

Implements the feature requested in issue #193: a client-callable HTTP
endpoint that gracefully closes alive MJPEG/stream connections server-side,
preventing lingering TCP sockets when UI components unmount.

Changes:
- StreamerInterface: add virtual stop() (no-op default) and get_client_id()
- StreamerBase: implement stop() (marks inactive, releases connection) and
  get_client_id(); populate client_id_ from 'client_id' query param
- WebVideoServer: register /shutdown route and implement handle_shutdown()
  which stops all streams for a topic, or a single named stream if
  client_id is provided; returns plain-text 'stopped=<count>'
- README: document /shutdown endpoint, parameters, and client_id usage

Rewrite of PR #194 adapted to the pluginlib-based architecture introduced
in #192 and the naming/style conventions enforced by #195.

varunverlencar commented:
@bjsowa made a clean rewrite, also curious to see how good claude was. lets see

bjsowa commented on src/web_video_server.cpp:
```diff
@@ -263,6 +266,49 @@ bool WebVideoServer::handle_stream_viewer(
+  if (verbose_) {
+    RCLCPP_INFO(
+      get_logger(), "Shutdown request for topic '%s' (client_id='%s'): stopped %d stream(s)",
+      topic.c_str(), client_id.c_str(), stopped);
```
This tries to log client_id even if it is empty.

bjsowa commented on include/web_video_server/streamer.hpp:
```diff
@@ -56,6 +56,11 @@ class StreamerInterface
+  /**
+   * @brief Stops the streaming process and marks the streamer as inactive.
+   */
+  virtual void stop() {}
```
This is an interface class, thus all methods should be pure virtual.
```suggestion
  virtual void stop() = 0;
```

bjsowa commented on include/web_video_server/streamer.hpp:
```diff
@@ -73,6 +78,14 @@ class StreamerInterface
+  virtual std::string get_client_id()
+  {
+    return "";
+  }
```
```suggestion
  virtual std::string get_client_id() = 0;
```

varunverlencar committed:
Update include/web_video_server/streamer.hpp

make pure virtual

Co-authored-by: Błażej Sowa <bsowa123@gmail.com>

varunverlencar committed:
Update include/web_video_server/streamer.hpp

make pure virtual

Co-authored-by: Błażej Sowa <bsowa123@gmail.com>

varunverlencar committed:
feat-193: fixed client_id empty logging

varunverlencar commented:
@bjsowa when is the next release date. whats the cadence for this project?

bjsowa commented:
> @bjsowa when is the next release date. whats the cadence for this project?

Probably when I get to review #198 and merge it. There is no strict release schedule.

If it is of importance to you, I can push a new release now.

varunverlencar commented:
> > @bjsowa when is the next release date. whats the cadence for this project?
> 
> Probably when I get to review #198 and merge it. There is no strict release schedule.
> 
> If it is of importance to you, I can push a new release now.

yes it would really help with an official release, appreciate it.

bjsowa commented:
I pushed a release. Should be available with the next ROS sync which should be soon for Jazzy, Kilted and Rolling and a little longer for Humble as it was synced recently.
